### ISSUE TITLE
make deep links shorter

### ISSUE THREAD

**Issue by ralfhauser** [mmmatthew/water-fountains.org#159]
The URL gets split on 2 pages

![image](https://user-images.githubusercontent.com/1681677/55326602-7236f300-5488-11e9-84ed-3297778f168d.png)

**Comment by ralfhauser** [comment]
1) we should offer on the detail view a button that automatically copies it into the clipboard
2) https://beta.water-fountains.org/zurich?lang=de&queryType=byId&database=wiki
data&idval=Q55170432  might be shortened to

https://water-fountains.org/zh?l=d&qt=id&db=wikidata&id=Q55170432 
https://water-fountains.org/zh?l=d&qt=id&db=wd&i=Q55170432 


a) eventually, we might do another try for a shorter domain URL again

**Comment by mmmatthew** [comment]
3) as discussed, we could also use the id to determine the database.

**Comment by ralfhauser** [comment]
Re 3) as all wikidate id's start with "Q" and no OSM does, please implement this as a quick-win (and make the "database" parameter optional for now as we currently only have 2 possible databases)

So we probably could get to
3a)   https://water-fountains.org/zh?l=d&i=Q55170432
which is quite unlikely to be hyphenated

**Comment by mmmatthew** [comment]
Is it too long to spell out zurich and lang? 

3b) https://beta.water-fountains.org/zurich?lang=de&idval=Q55170432

this 58 chars, and the maximum allowable before getting split (according to first image) is 76 chars, so we still have room for long location names.

**Comment by ralfhauser** [comment]
> Is it too long to spell out zurich and lang?

No, it is not.

But if we touch it, I would allow for the shorter variant in parallel

------
and what is the rationale to make it as long as "idval" and not just "id" or even "i" ?

**Comment by mmmatthew** [comment]
I am unhappy with "idval" but "id" seemed to be used internally as it caused the app to crash when I would use it.

**Comment by ralfhauser** [comment]
then pls go vor "i" that is even shorter

**Comment by ralfhauser** [comment]
4) as for the city-Kürzel:
a) a for pioneers, we can stay with just 2 letters like "ZH".
b) for internationalization, I suggest to take b1) the ISO 2-Letter-Country id and then b2) a country-internal 2-char identifier

==>  "CH-ZH"

Re b2) that should be semantically taken out of wikidata

**Comment by mmmatthew** [comment]
Here is the new format:

https://beta.water-fountains.org/ch-zh?l=de&i=Q55169550

**Comment by mmmatthew** [comment]
The old urls should still be compatible.

**Comment by ralfhauser** [comment]
The language in https://beta.water-fountains.org/us-nyc?l=en does not work   :(

5) could the key languages also be single-letter ? d/f/i/e ?
so, we would get to https://beta.water-fountains.org/us-ny?l=e

**Comment by ralfhauser** [comment]
The URLs in 2 and 3 don't work.

Is 4b2 implemented ?

**Comment by mmmatthew** [comment]
I only implemented the shortest variant, not the intermediate ideas. That is why I didn't implement the variants 2 and 3 b

3a  works when you include the beta before.

4b2 is a new issue

**Comment by ralfhauser** [comment]
> 3a works when you include the beta before.

Will it work without "beta" once you released production ?

**Comment by mmmatthew** [comment]
I don't see why not.

**Comment by ralfhauser** [comment]
> I don't see why not.

Then, let's verify and close this once we are there

**Comment by mmmatthew** [comment]
> Then, let's verify and close this once we are there

https://water-fountains.org/zh?l=d&i=Q55170432 works

**Comment by ralfhauser** [comment]
Thanks - the original idea of #35 or rather #244 is still pending ?

**Comment by mmmatthew** [comment]
#244 is still pending. Fountains are aggregated and stored on a city-by-city basis, so if only an identifier is provided, an extra step is required before fetching the data, to search city databases for specific fountain. In the long term, it makes sense to upgrade how the aggregated fountains are stored and use a database instead of the current simple cache solution.

**Comment by ralfhauser** [comment]
I guess the "extra step" would be to create on the bi-hourly region-reload an index of all Q-Ids-vs. City.

But since we have < 10'000 fountains covered, such an index should be quick to be built in no time and querying it should be even quicker

**Comment by ralfhauser** [comment]
see also https://github.com/water-fountains/import2wikidata/issues/10
